### ISSUE TITLE
LaBSE Sentence Embeddings model output vectors are not equal to original

### ISSUE THREAD

**Issue by Fikavec** [JohnSnowLabs/spark-nlp#2846]
I’am using embeddings from example https://nlp.johnsnowlabs.com/2020/09/23/labse.html and output vectors although close, but not equal to original vectors https://tfhub.dev/google/LaBSE/1 Why? 
How original vectors converted to this model, maybe original model modified or finetuned or spark-nlp using different normalization? - on example multiclass/multilabel tasks spark-nlp LaBSE embeddings work differently than original vectors, how get this vectors from original model? Original vectors are case sensitive, but in spark-nlp config this vectors are case insensitive, original vectors max_seq_length = 64, but in spark-nlp = 128... any other differences? For re-produce problem: run code from original vectors page with  ["I love NLP", "Many thanks"] and compare outputs to outputs from spark-nlp model page. 
How open spark-nlp model by tf.saved_model.load after unzip pb file and run infer in tensorflow for "low-level" compare outputs and analyse problem (tensorflow 2.x can't load unzipped bert_sentence_tensorflow model) - maybe difference/problem occurs in converted tf model? 

![difference](https://user-images.githubusercontent.com/83672821/117535864-712a0580-b000-11eb-82b4-cdad4be89831.png)

Notebook to reproduce: 
[Compare_outputs_of_Spark_nlp_LaBSE_embeddings_and_Original_TF_hub_LaBSE_embeddings.zip](https://github.com/JohnSnowLabs/spark-nlp/files/6445681/Compare_outputs_of_Spark_nlp_LaBSE_embeddings_and_Original_TF_hub_LaBSE_embeddings.zip)

**Comment by Fikavec** [comment]
We can see differences in embeddings on examples which mean the same thing from Spark-nlp 20 langs multilingual classification: [https://github.com/JohnSnowLabs/nlu/blob/master/examples/colab/Training/multi_lingual/multi_class_text_classification/NLU_training_multi_lingual_multi_class_text_classifier_demo.ipynb](url)

1. Science has advanced rapidly over the last century 
2. Die Wissenschaft hat im letzten Jahrhundert rasante Fortschritte gemacht 
3. 在上个世纪，科学发展迅速 
4. Die wetenskap het die afgelope eeu vinnig gevorder 
5. Khoa học đã phát triển nhanh chóng trong thế kỷ qua 
6. 科学は前世紀にわたって急速に進歩しました 
7. Isayensi ithuthuke ngokushesha ngekhulu leminyaka elidlule 
8. Bilim, geçen yüzyılda hızla ilerledi 
9. המדע התקדם במהירות במהלך המאה האחרונה 
10. గత శతాబ్దంలో సైన్స్ వేగంగా అభివృద్ధి చెందింది 
11. Наука стремительно развивалась за последнее столетие 
12. سائنس گذشتہ صدی کے دوران تیزی سے ترقی کرچکی ہے 
13. "विज्ञान पिछली सदी में तेजी से आगे बढ़ा है 
14. Соңгы гасырда фән тиз үсә 
15. La science a progressé rapidement au cours du siècle dernier 
16. วิทยาศาสตร์ก้าวหน้าอย่างรวดเร็วในช่วงศตวรรษที่ผ่านมา 
17. វិទ្យាសាស្ត្របានជឿនលឿនយ៉ាងលឿនក្នុងរយៈពេលមួយសតវត្សចុងក្រោយនេះ 
18. וויסנשאַפֿט איז ראַפּאַדלי אַוואַנסירטע איבער די לעצטע יאָרהונדערט 
19. Илим акыркы кылымда тездик менен өнүккөн 
20. கடந்த நூற்றாண்டில் அறிவியல் வேகமாக முன்னேறியுள்ளது 

**Spark-nlp LaBSE embeddings cosine similiraty heatmap**
![spark_cossim](https://user-images.githubusercontent.com/83672821/117530412-ba209080-afe5-11eb-9507-b3e810b1deb9.png)
**Original TF-Hub LaBSE embeddings cosine similiraty heatmap with do_lower_case = False and max_seq_length = 128**
![labse_lowercase_cossim](https://user-images.githubusercontent.com/83672821/117530918-7a0edd00-afe8-11eb-9606-6a7cbb0ed534.png)
**Original TF-Hub LaBSE embeddings cosine similiraty heatmap with do_lower_case = True and max_seq_length = 128**
![labse_cossim](https://user-images.githubusercontent.com/83672821/117530936-8eeb7080-afe8-11eb-8fcb-2ebac48c340d.png)

**Comment by Fikavec** [comment]
For simplify debugging issue there are attahed csv with sorted sentences by distance from Spark-nlp embeddings to Original Tf-hub model and colab for reproduce:
[compute_sentence_difference_distance.zip](https://github.com/JohnSnowLabs/spark-nlp/files/6446205/compute_sentence_difference_distance.zip)
![diff](https://user-images.githubusercontent.com/83672821/117547103-7e61e700-b036-11eb-8476-382c09b753f3.png)

**Comment by Fikavec** [comment]
With activated tf.float64 and setting max_seq_length = 128 and lowercased input sentences to Original TF-Hub model (! lowercased input sentences, but don't touch do_lower_case option in original model example code, because after setting do_lower_case=True this model makes much worse embeddings than lowercased input) on tasks from 
[multi_class_text_classification](https://github.com/JohnSnowLabs/nlu/tree/master/examples/colab/Training/multi_lingual/multi_class_text_classification) and test examples from sections "Model with 20 languages!" model can achieve (given results from run to run is not stable because train_test_split, without setting random sid, model iitialization etc) the following results: 

- **Amazon**
 ![Amazon](https://user-images.githubusercontent.com/83672821/117548121-d2bb9580-b03b-11eb-9954-ba80a235a499.png)
 In Spark-nlp example:
 ![spark_amazon](https://user-images.githubusercontent.com/83672821/117548126-db13d080-b03b-11eb-9177-70dd03248a4a.png)
 
- **Tripadvice**
 ![tripadviser](https://user-images.githubusercontent.com/83672821/117548135-e5ce6580-b03b-11eb-9552-b879e3ce869f.png)
 In Spark-nlp example:
 ![spark_tripadviser](https://user-images.githubusercontent.com/83672821/117548140-ecf57380-b03b-11eb-9fae-e49a42660d03.png)
- **News**
 ![News](https://user-images.githubusercontent.com/83672821/117548388-68a3f000-b03d-11eb-9035-a0b10e9b082e.png)
 In Spark-nlp example:
 ![spark_news](https://user-images.githubusercontent.com/83672821/117548174-2b8b2e00-b03c-11eb-9598-d3ea4511a8e0.png)

Maybe main diferrence from Original model in Spark-nlp multilingual tokenization procedure and embeddings difference appears ony when input sentences are on some languages: Vietnamese, Japan, Urdu... ?

**Comment by maziyarpanahi** [comment]
These are very interesting findings, thanks for sharing them. A couple of things come to my mind:

Some explainations:

- The case sensitivity in Spark NLP is defined during runtime, depending on the param being set by the user it will lowercase already tokenized text or leave it as it is. Then it tries to tokenize the tokens by BERT tokenizer to encode the piece ids. So you can change it depending on the model if it supports cased or uncased. (you can test both these scenarios)
- Same goes for `max_seq_length`, you can change that during runtime as well in Spark NLP via `setMaxSentenceLength`. The BERT models support up to 512 so you can choose the length depedning on your needs, like matching two comparisons.
- The actual vectors (floats) can be different from the original since there is a conversion and change of environment. However, this should change much in the actual end results
- We also didn't change anything in the original model as far as I know, we just simple used saved_model and then loaded and saved it in Spark NLP without any change.

Some strong possibilities:

- What you have suspected coould one of the reasons or the reason for this inequality. If you can do your F1 based on each language it may reveal that our custom BERT tokenizer may not do as good on some languages as the orignal BERT tokenizer from TF Hub. (which if this is the case we will fix it)
- Another possibility is that unlike normal use of BERT where the string goes to the tokenizer, breaks into word peices, then they are encoded to piece ids, etc. we have to fiest toeknize the string via Tokenizer, then pass those tokens to BERT Tokenizer, breaks to word peices, then piece ids, etc. and then the vectors are map back to the original tokens (custom). All the NLP tasks in Spark NLP rely on the Tokenizer so we must align every custom tokenization via BERT or BPE or SentencePiece back to the orignal so it can be used in Classification or NER. This may cause different vectors because of what goes before or after a token etc.

I suspect +2/-2 up and down F1-score in English is caused by the custom Tokenization and the big difference in the F1-score within multi-lingual is cased by a bad multi-lingual tokenization (to be confirmed if you can do the metrics by language to compare how good and bad they are compare to Spark NLP). 

Will continue debugging since this is a very useful multi-lingual embeddings.

**Comment by maziyarpanahi** [comment]
just tagging you @C-K-Loan as an FYI to be sure `LaBSE` bad performance is due to bad BERT multi-lingual tokenization. (we always used the English so we might have never noticed)

**Comment by Fikavec** [comment]
Thanks for your reply. For this reason
> to be confirmed if you can do the metrics by language to compare how good and bad they are compared to Spark NLP

in order to avoid such hard-to-find differences / inaccuracies in the process of integrating multilingual models in the future and to assess the quality of multilingual embeddings, I propose two methods:
1) Carefully prepare a few groups of sentences which mean the same thing in 100+ languages (sentences about the same thing should be simple and unambiguous without modern vocabulary, hand-picked from texts or from native speakers, but not from an automatic translator). For each group draw cosine heatmap for visual compare [as i do here](https://github.com/JohnSnowLabs/spark-nlp/issues/2846#issuecomment-835179008). Heatmaps of original and integrated models can be visual compared to detect differences in some languages. At the same time in a quality multi-lingual embeddings for simple unambiguous thing all squares should be closer to 1 (and lighter). 
2) Make automated tests for compute distances from original models to integrated on sentences in 100+ langs (supported by integrated models) like 1) [as i do here ](https://github.com/JohnSnowLabs/spark-nlp/issues/2846#issuecomment-835427327)

I think these techniques will be useful not only for solving this issue with LaBSE, but for integrating any multilingual models in the future (like USE, LASER, Sentence Transformers models etc) as a simple fast standard tests of the correctness of integration process.
All F1 [presented here](https://github.com/JohnSnowLabs/spark-nlp/issues/2846#issuecomment-835445175) i'm get for sentences at 20 languages (not in english sentences) from datasets - test_df[["test_sentences","y"]].iloc[:100].

**Comment by Fikavec** [comment]
For test examples from section "Model with 20 languages!" from [2 Class Stock Market Sentiment Classifier Training](https://github.com/JohnSnowLabs/nlu/blob/master/examples/colab/Training/multi_lingual/binary_text_classification/NLU_multi_lingual_training_sentiment_classifier_demo_stock_market.ipynb) i got the following results with original tf-hub model embeddings (+max_seq_length = 128):
![stock](https://user-images.githubusercontent.com/83672821/117581810-673efa00-b107-11eb-9bd1-14d491d0b10c.png)
In Spark-nlp example:
![stock_spark](https://user-images.githubusercontent.com/83672821/117581842-9a818900-b107-11eb-8f5c-c60af854a5b0.png)

**Comment by maziyarpanahi** [comment]
Thanks, we will investigate this further. The hg difference between macro and the micro in Spark NLP indicates some classes did good some very bad. Which means the tokenizer may not be doing so great on some languages.
That’s being said, some languages don’t have any whitespace and require segmentations such as Chinese, Korean, Japanese, etc. For those we have WordSegmenter annotator which is also one of the issue that these languages are not being tokenized at all by not using WordSegmenter instead of a normal Tokenizer.


@C-K-Loan Could you please do the test by separating those languages and do them in two different groups? (Some with Tokenizer and the rest by using WordSegmenter)

**Comment by Fikavec** [comment]
This option is very usefull
> The case sensitivity in Spark NLP is defined during runtime, depending on the param being set by the user it will lowercase already tokenized text or leave it as it is. Then it tries to tokenize the tokens by BERT tokenizer to encode the piece ids. So you can change it depending on the model if it supports cased or uncased. (you can test both these scenarios)

but as I understand, by default in Spark-nlp model config this model is **not** caseSensitive:

> {"class":"com.johnsnowlabs.nlp.embeddings.BertSentenceEmbeddings",
> 	"timestamp":1600858075694,
> 	"sparkVersion":"2.4.4",
> 	"uid":"BERT_SENTENCE_EMBEDDINGS_f12be4ed4fef",
> 	"paramMap":{
> 		"storageRef":"labse",
> 		"inputCols":["sentence"],
> 		**"caseSensitive":false,**
> 		"dimension":768,
> 		"outputCol":"bert"},
> 	"defaultParamMap":{
> 		"maxSentenceLength":128,
> 		"storageRef":"BERT_SENTENCE_EMBEDDINGS_f12be4ed4fef",
> 		"lazyAnnotator":false,
> 		**"caseSensitive":false,**
> 		"batchSize":32,"
> 		dimension":768}
> }

As I know 

- the case sensitivity is the advantage of the model
- In many cases, if a model has learned to be case-sensitive, she may not know these lowercase words and artificially predict their vectors (Greetings, Personal Pronoun, Proper Names, Nationalities, Сountries, States, Towns, Days Of The Week, Months, Holidays And School Subjects, The Names Of Universities And Schools, Abriviations etc)

And I'm get testing results that in this model same thing sentences in many languages without lowercased are closer to each other and distance between sentence and lowercased sentense sometimes not so small:

![lowercase](https://user-images.githubusercontent.com/83672821/117663061-e89a9900-b1a8-11eb-996f-dd5ef51d94aa.PNG)

I'm think that caseSensitive:false by default for this model may get better results on small train/test datasets (if so, then it can be shown separately in the examples of using this model), but in general, this may not be so good, as it disables the advantage of the original model and may lead to the above inaccuracies, or am I wrong?

**Comment by maziyarpanahi** [comment]
That’s true, I think by mistake that param was set to false when it was saved. Since the current version is 1 for TF v1, we will make a new one for TF v2 from the version 2 and make sure that param is set to true as well when we upload it.

**Comment by Fikavec** [comment]
With a powerfull of [wikipedia interlanguage links](https://simple.wikipedia.org/wiki/Wikipedia:List_of_articles_all_languages_should_have) for future I'm prepared multilingual test with sentences about one [thing](https://en.wikipedia.org/wiki/World_Health_Organization) on 151 languages - 88 languages is supported by LaBSE (I think that LaBSE supports more then 88 languages of 151, but LaBSE languages ISO 639-1/639-2 two code not in wikipedia languages code list -  to do: needs manually align).  

> Colab notebook to reproduce and heatmaps png:
> [visual_multilingual_embeddings_compare_colab.zip](https://github.com/JohnSnowLabs/spark-nlp/files/6454646/visual_multilingual_embeddings_compare_colab.zip)

*Test sentence language is indicated at the top of the images
**Original TF-Hub LaBSE model cosine distance between sentences about one thing on 88 languages heatmap**
![Original_TF-HUB_LaBSE_88langs_same_thing_embeddings](https://user-images.githubusercontent.com/83672821/117716021-bd806b80-b1e1-11eb-8f70-fa87fcfae2e4.png)
**SparkNLP LaBSE model cosine distance between sentences about one thing on 88 languages heatmap**
![SparkNLP_88langs_same_thing_embeddings](https://user-images.githubusercontent.com/83672821/117716109-dd179400-b1e1-11eb-919d-7a25c6409bab.png)

**Comment by Fikavec** [comment]
Updated the above [multilingual dataset with code](https://github.com/JohnSnowLabs/spark-nlp/issues/2846#issuecomment-837257841) for debugging and [idea](https://github.com/JohnSnowLabs/spark-nlp/issues/2846#issuecomment-835846740) for future autotesting. Thank you for great job on this amazing project and many interesting workshops. Looking forward [Multilingual T5 (mT5)](https://github.com/google-research/multilingual-t5) and future releases.

**Comment by Fikavec** [comment]
Facebook AI Open-Source [The FLORES-101 Data Set](https://github.com/facebookresearch/flores) consisting of 3001 sentences translated in 101 languages by professional translators ([more info](https://ai.facebook.com/blog/the-flores-101-data-set-helping-build-better-translation-systems-around-the-world)). Maybe it is great dataset to measure quality of multilingual embeddings and starting point to create auto tests as I’m describes early.
